plugins/pipewire/PipeWireDevice.cpp

```diff
@@ -0,0 +1,300 @@
+/*******************************************************************************
+ * Copyright 2009-2024 Jörg Müller
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *	 http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ ******************************************************************************/
+
+#include "PipeWireDevice.h"
+
+#include <spa/param/audio/format-utils.h>
+
+#include "Exception.h"
+#include "IReader.h"
+#include "PipeWireLibrary.h"
+
+#include "devices/DeviceManager.h"
+#include "devices/IDeviceFactory.h"
+
+AUD_NAMESPACE_BEGIN
+
+PipeWireDevice::PipeWireSynchronizer::PipeWireSynchronizer(PipeWireDevice* device) : m_device(device)
+{
+}
+
+void PipeWireDevice::PipeWireSynchronizer::update_tick_start()
+{
+	pw_time tm;
+	AUD_pw_stream_get_time_n(m_device->m_stream, &tm, sizeof(tm));
+	m_tick_start = tm.ticks;
+	m_seek_pos = m_timeline_pos;
+}
+
+void PipeWireDevice::PipeWireSynchronizer::seek(std::shared_ptr<IHandle> handle, double time)
+{
+	pw_time tm;
+	AUD_pw_stream_get_time_n(m_device->m_stream, &tm, sizeof(tm));
+	m_tick_start = tm.ticks;
+	m_seek_pos = m_timeline_pos = time;
+	handle->seek(time);
+}
+
+double PipeWireDevice::PipeWireSynchronizer::getPosition(std::shared_ptr<IHandle> handle)
+{
+	pw_time tm;
+	AUD_pw_stream_get_time_n(m_device->m_stream, &tm, sizeof(tm));
+	uint64_t now = AUD_pw_stream_get_nsec(m_device->m_stream);
+	int64_t diff = now - tm.now;
+	int64_t elapsed = (tm.rate.denom * diff) / (tm.rate.num * SPA_NSEC_PER_SEC);
+
+	/* Calculate the elapsed time in seconds from the last seek position. */
+	double elapsed_time = (tm.ticks - m_tick_start + elapsed) * tm.rate.num / double(tm.rate.denom);
+	/* Convert to seconds and add the last seek position timestamp. */
+	m_timeline_pos = elapsed_time + m_seek_pos;
+	return m_timeline_pos;
+}
+
+void PipeWireDevice::handle_state_changed(void* device_ptr, enum pw_stream_state old, enum pw_stream_state state, const char* error)
+{
+	PipeWireDevice* device = (PipeWireDevice*) device_ptr;
+	// fprintf(stderr, "stream state: \"%s\"\n", pw_stream_state_as_string(state));
+	if(state == PW_STREAM_STATE_STREAMING)
+	{
+		/* When we activate/unpause the stream, we need to make sure that
+		 * we update the refrence tick number as the ticks we get from PipeWire
+		 * might have continued to count up after we paused the stream.
+		 */
+		device->m_synchronizer.update_tick_start();
+	}
+}
+
+void PipeWireDevice::mix_audio_buffer(void* device_ptr)
```

Comment:
Like other devices which have such a callback, it would be much better to have a lightweight and lockless callback that returns as quickly as possible with mixing actually happening in a separate thread. Basically like Jack and pulseaudio are doing it.

Reply:
This might explain why I got higher reported latency in `pw-top` with this backend than if I were using the Jack or the Pulseaudio backend... :sweat_smile: 

Thanks for pointing this out!